### PR TITLE
Bidirectional

### PR DESCRIPTION
**Pull request by Lou1sM**
Added a second prediction head that predicts the end explicitly, and then regresses back to the start--so the mirror image of the existing prediction head. Both predictions are then fused, taking the start time from the former and the end time from the latter, and add in some high probability predictions from both as well. 

Sorry there are so many lines marked as changed. Most of them are because I removed trailing whitespace when my linter kept prompting me to. The substantive changes are in

- voxaboxen/data/data.py in DetectionDataset.get_annotation() to create the reversed data, 
- voxaboxen/model/model.py to add the reverse predictions,
- voxaboxen/training/train.py to train on both the forward and backward predictions,
- voxaboxen/evaluation/evaluation.py to generate the backward predictions and match them with the forward predictions,
- voxaboxen/training/train_model.py to fit the threshold for 'high probability predictions' on the val set.

### CODE REVIEW

**Comment by benjaminsshoffman** [comment]
yoo this is awesome! i've been wanting to do something like this but haven't found the time. excited to try it out, i'll review and then hopefully merge in

**Comment by Lou1sM** [comment]
Great! Let me know if you have any questions or there's anything I can help with

File: voxaboxen/training/train_model.py [hunk 1]
```diff
@@ -8,41 +10,70 @@
...
   ## Setup
   args = parse_args(args)
 
   set_seed(args.seed)
-  
+
   experiment_dir = os.path.join(args.project_dir, args.name)
   setattr(args, 'experiment_dir', str(experiment_dir))
-  if not os.path.exists(args.experiment_dir):
+  if os.path.exists(args.experiment_dir) and args.overwrite and args.name!='demo':
```
Comment (benjaminsshoffman) [inline voxaboxen/training/train_model.py:27]: should be `not args.overwrite`?

File: voxaboxen/training/train_model.py [hunk 2]
```diff
@@ -8,41 +10,70 @@
...
-  
+
   save_params(args)
   if hasattr(args,'stereo') and args.stereo:
     model = DetectionModelStereo(args)
   else:
     model = DetectionModel(args)
-  
+
+  if args.reload_from is not None:
```
Comment (benjaminsshoffman) [inline voxaboxen/training/train_model.py:42]: I think this is redundant with `args.previous_checkpoint_fp`, which currently happens in `train.py`

**Comment by benjaminsshoffman** [comment]
Ok this looks generally good, I had a few small comments I put into the code.

My main request, which will require a little bit more work, is to ensure backward compatibility. I have a lot of trained models I'm using for ongoing projects, and I don't think i can use them under this revision. I think this would involve making your forward-backward model a subclass of the current detection model (similar to how stereo detection model is), and modifying the training & evaluation setup to allow for both situations. It's a bit annoying, sorry about that.

A second request would be to incorporate the forward-backward model in the inference script -- ideally you'd have a flag to specify that you want matched or combined predictions, and it would generate predictions of just that type. I think (maybe i'm wrong) that in the current setup it would not generate matched predictions when running inference.

let me know what you think. thanks for all your work on this, it's so cool to see it!

**Comment by Lou1sM** [comment]
Hey, thanks for the comments. Yeah, no problem to refactor into a subclass. Do we want to consider the possibility of combining multiple additions to the base model, i.e. be able to use any subset of stereo, fwd-bck and whatever else you're working on? In that case, separate subclasses for each could make it difficult, but I don't know if this is a consideration.

Re the second request, currently evaluate_based_on_manifest always generates 4 types of predictions, and returns a dict with keys fwd, bck, comb and match. This is what's called in voxaboxen/training/train_model.py. Is this the inference script you mean to modify? There could be an argparse arg to pick one of comb and match, and only print and save the results for that, or we could just always keep comb as that seems to give better results.

File: voxaboxen/training/train_model.py (hunk 1 above)
Comment (Lou1sM) [reply voxaboxen/training/train_model.py:27]: err yes, will fix this

File: voxaboxen/training/train_model.py (hunk 2 above)
Comment (Lou1sM) [reply voxaboxen/training/train_model.py:42]: oh yes I see, I added this for when n_epochs==0 (i.e. just reloading and testing some checkpoint), but yeah clumsy to have two different ways of loading checkpoints, could just make a tiny change to train.py, line 122: 

`val_epoch(model, t+1, val_dataloader, args)` ---> `val_epoch(model, args.n_epochs, val_dataloader, args)`?

**Comment by benjaminsshoffman** [comment]
Ok yeah I see what you mean about the subclassing. Maybe another approach, which I think would also be backward compatible, would be to have some conditional statements in the `__init__` and `forward` for `DetectionModel`. like `if bidirectional in args and args.bidirectional: self.rev_detection_head = DetectionHead(args, embedding_dim = embedding_dim)`, etc. Similar conditional statements would need to get added during training, eval, and for subclasses like Stereo. 

FYI I also have a segmentation option (i.e. just doing multi-label classification at 50Hz), which I haven't quite found time to finish and commit, but trying to think about how to integrate multiple options like these into training and evaluation, without totally breaking stuff. If you have another idea, i'm open to it.

For the second inference comment, I was referring to `voxaboxen/inference/inference.py`, since that's the piece used to apply the model to a bunch of unlabeled audio. I think that the matched predictions don't get generated there?

**Comment by Lou1sM** [comment]
Yeah flags for each of the different potential add-ons could work, maybe as model properties--so one big class with conditionals in a lot of the methods. I think some version of that is the only way to go if we want to be able to combine different options. The methods will be longer, but the experience of using the class could still up pretty clean (HuggingFace code is kinda like this). 

Does the segmentation option mean identifying different components in a vocalization, like in speech segmentation? That may still be compatible with bidirectional, but if not there could be some checks like `if args.bidirectional and args.segmentations: raise ValueError`.

Oh yes, I haven't touched anything in `voxaboxen/inference/`. ATM, fitting the threshold on the val set and applying it to the test set happens in `train_model.py`. I could just save this fit threshold as a model property and then use that in  `voxaboxen/inference/`?

**Comment by benjaminsshoffman** [comment]
Ok sounds good, let's do flags + conditional statements in one big class. Feel free to take a stab at it!

By segmentation I just mean making a prediction for each frame (where frames are at 50Hz), and then consecutive frames with the same prediction are combined if they're the same. Like semantic segmentation for images (but multi-label to allow overlapping vox). This is for a situation where we have some baby birds making lots of begging sounds, and it's not clear even to the listener when one sound starts and the other ends. I think it would not be compatible with bidirectional as it currently is (it does not use the detection or regression outputs), so raising an error seems appropriate.

Saving as a model property sounds good! would still like a manual override, even if just for debugging, but loading it eg with the model checkpoint would make things easy.

**Comment by Lou1sM** [comment]
Ah I was wondering if you'd tried this frame-wise prediction approach, I expect there are advantages and disadvantages. Yeah I think not compatible with bidirectional atm, could maybe be combined as some ensemble of the different types of predictions, but that would be a future thing.

Great, I'll work on both these things, hope to find some time at the end of next week.

**Comment by Lou1sM** [comment]
Hey, I had a go at these two points.

For the model flags, I think the only two that are options now are bidirectional and stereo, I also included an arg for segmentation but it doesn't do anything other than raise an error if passed with bidirectional. I tried using the stereo option but it gives an error that the data is the wrong shape, which is probably expected. When is stereo supposed to work? I think the DetectionModelStereo can be deleted, but left it there for now.

For the inference script, saving as a parameter works fine, the only slightly awkward bit being the need to convert it to a float with `.item()` before passing to `combine_fwd_bck_preds`. I included an inference arg `disable-bidirectional` to use as a regular model even if it was trained as a bidirectional one.

**Comment by benjaminsshoffman** [comment]
Hey i'm sorry it's taking me so long to get to this -- had a conference and some other things come up. haven't forgotten it!

**Pull request merged**